### PR TITLE
[IAP] Draft of new upgrades screen

### PR DESCRIPTION
**Pull request by joshheald**
Merge after: #10165 

## Description

This needs some polish before merging, if we do want to merge it at all. I think we should move away from List in favour of a scrollable VStack, because of some strange UI glitches.

Also we might consider a custom picker control for the cards – at the moment the tap gesture only works if you tap on some text in the section, not the white space. However, a picker may not work well with the design direction, which may have an expandable features part of each section.

## Screenshots

https://github.com/woocommerce/woocommerce-ios/assets/2472348/0579de09-c946-4253-9db5-e8455e3dc96b

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Upgrades/OwnerUpgradesView.swift [hunk 1]
```diff
@@ -1,75 +1,111 @@
...
 
-                    VStack(alignment: .leading) {
-                        Text(upgradePlan.wooPlan.shortName)
-                            .font(.largeTitle)
-                            .accessibilityAddTraits(.isHeader)
-                        Text(upgradePlan.wooPlan.planDescription)
-                            .font(.subheadline)
-                    }
+    @State private var paymentFrequency: WooPlan.PlanFrequency = .year
+    private var paymentFrequencies: [WooPlan.PlanFrequency] = [.year, .month]
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Upgrades/OwnerUpgradesView.swift:22]: At some point in the future we should make these available to the view through a viewmodel, and not expose Yosemite directly, thoughts? We currently don't have anything driving the view's data, just the closures passing data back to the `UpgradesView` when needed

File: WooCommerce/Classes/ViewRelated/Upgrades/OwnerUpgradesView.swift [hunk 2]
```diff
@@ -1,75 +1,111 @@
...
 
 struct OwnerUpgradesView: View {
-    @State var upgradePlan: WooWPComPlan
-    @State var isPurchasing = false
-    let purchasePlanAction: () -> Void
-    @State var isLoading: Bool = false
+    @State var upgradePlans: [WooWPComPlan]
+    @State var isPurchasing: Bool
+    let purchasePlanAction: (WooWPComPlan) -> Void
+    @State var isLoading: Bool
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Upgrades/OwnerUpgradesView.swift:9]: The properties marked with underscore went a bit over my head here. Is this the same as providing initial values as follows?

```suggestion
    @State var upgradePlans: [WooWPComPlan] = []
    @State var isPurchasing: Bool = false
    let purchasePlanAction: (WooWPComPlan) -> Void
    @State var isLoading: Bool = false
```

File: WooCommerce/Classes/ViewModels/InAppPurchases/UpgradesViewModel.swift [hunk 3]
```diff
@@ -70,13 +72,13 @@ final class UpgradesViewModel: ObservableObject {
                 throw PrePurchaseError.maximumSitesUpgraded
             }
 
-            guard let plan = try await retrievePlanDetailsIfAvailable(.essentialMonthly,
-                                                                      from: wpcomPlans,
-                                                                      hardcodedPlanDataIsValid: hardcodedPlanDataIsValid)
-            else {
+            plans = try await retrievePlanDetailsIfAvailable([.essentialMonthly, .essentialYearly, .performanceMonthly, .performanceYearly],
+                                                                 from: wpcomPlans,
+                                                                 hardcodedPlanDataIsValid: hardcodedPlanDataIsValid)
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewModels/InAppPurchases/UpgradesViewModel.swift:77]: Do we want to hide the additional plans behind a flag at this point? Sort of:
```suggestion
var availablePlans = [AvailableInAppPurchasesWPComPlans]()
if ServiceLocator.featureFlagService.isFeatureFlagEnabled(.freeTrialInAppPurchasesUpgradeM2) {
  availablePlans = [.essentialMonthly, .essentialYearly, .performanceMonthly, .performanceYearly]
} else {
  availablePlans = [.essentialMonthly]
}
plans = try await retrievePlanDetailsIfAvailable(availablePlans, from: wpcomPlans, hardcodedPlanDataIsValid: hardcodedPlanDataIsValid)
```

File: WooCommerce/Classes/ViewModels/InAppPurchases/UpgradesViewModel.swift [hunk 4]
```diff
@@ -92,11 +94,11 @@ final class UpgradesViewModel: ObservableObject {
     @MainActor
     func purchasePlan(with planID: String) async {
         analytics.track(event: .InAppPurchases.planUpgradePurchaseButtonTapped(planID))
-        guard let wooWPComPlan = planCanBePurchasedFromCurrentState() else {
+        guard let wooWPComPlan = planCanBePurchasedFromCurrentState(with: planID) else {
             return
         }
 
-        upgradeViewState = .purchasing(wooWPComPlan)
+        upgradeViewState = .purchasing(wooWPComPlan, plans)
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewModels/InAppPurchases/UpgradesViewModel.swift:101]: Just leaving a note here for reference: I haven't tested this fully, and not sure if we would need it, but if we happen to not want both associated values to the .purchasing state we could also retrieve `plans` from the `planCanBePurchasedFromCurrentState` method above if we return a tuple, sort of:

```
func planCanBePurchasedFromCurrentState(with planID: String) -> (WooWPComPlan?, [WooWPComPlan])? {
    switch upgradeViewState {
    case .loaded(let plans):
        let selectedPlan = plans.first(where: { $0.wpComPlan.id == planID })
        return (selectedPlan, plans)
    case .purchaseUpgradeError(.inAppPurchaseFailed(let plan, _)):
        return (plan, [])
    default:
        return nil
    }
}
```

File: Yosemite/Yosemite/Model/WooPlans/WooPlan.swift [hunk 5]
```diff
@@ -276,15 +330,35 @@ private extension WooPlan {
     }
 
     enum Localization {
-        static let essentialPlanName = NSLocalizedString(
-            "Woo Express Essential Monthly",
-            comment: "Title for the plan on a screen showing the Woo Express Essential Monthly plan upgrade benefits")
+        static let essentialPlanNameFormat = NSLocalizedString(
+            "Woo Express Essential %1$@",
```
Comment (iamgabrielma) [inline Yosemite/Yosemite/Model/WooPlans/WooPlan.swift:334]: 🤔 In this string and similar others, should these be `NSLocalizedStrings`? Or we want the name to be `Woo Express Essential monthly/yearly` in all languages? Or perhaps translate only the monthly/yearly part?

**Review by iamgabrielma — APPROVED** [review]
Looking great! 🚢 

I know this is a draft, but just noting that there's a glitch when switching frequency and attempting a purchase, so we can keep a look on it: 

Select "Pay Monthly" > Select any of the plans > Tap "Purchase Essential Monthly" > See the background switching to annual plans (the Apple modal is still selecting the correct plan)

File: WooCommerce/Classes/ViewModels/InAppPurchases/UpgradesViewModel.swift (hunk 3 above)
Comment (joshheald) [reply WooCommerce/Classes/ViewModels/InAppPurchases/UpgradesViewModel.swift:77]: I don't think so; the whole view and view model is hidden behind that flag, so we can keep it simple.

File: WooCommerce/Classes/ViewRelated/Upgrades/OwnerUpgradesView.swift (hunk 1 above)
Comment (joshheald) [reply WooCommerce/Classes/ViewRelated/Upgrades/OwnerUpgradesView.swift:22]: The view using Yosemite's types is fine... there's no reason that we'd need to wrap them in one part of the WooCommerce App target to hide them from another part of the same target.

Putting them on the viewModel may still be desirable, particularly the `paymentFrequencies` array, we'll see how we go with that. For this, I felt that while the user is making their selection, it's all just private view state, and once they press the button, the view model can be notified to handle the purchase.

File: WooCommerce/Classes/ViewRelated/Upgrades/OwnerUpgradesView.swift (hunk 2 above)
Comment (joshheald) [reply WooCommerce/Classes/ViewRelated/Upgrades/OwnerUpgradesView.swift:9]: Yeah, it's just that when we do it in an init, we need to make the `State` with an initialValue. With the state values being private, we don't have a free init method for them, so this becomes required if we want to be able to set them when we make the view (which we do)

File: WooCommerce/Classes/ViewRelated/Upgrades/OwnerUpgradesView.swift (hunk 1 above)
Comment (joshheald) [reply WooCommerce/Classes/ViewRelated/Upgrades/OwnerUpgradesView.swift:22]: `_upgradePlans` references the underlying property, without going through the `@State` property wrapper, which doesn't exist until we call `State.init(initialValue:)`

File: Yosemite/Yosemite/Model/WooPlans/WooPlan.swift (hunk 5 above)
Comment (joshheald) [reply Yosemite/Yosemite/Model/WooPlans/WooPlan.swift:334]: I'm pretty sure we want to at least offer a translation option. What does the web do in other locales?

Currently, we translate all of the name.

**Pull request merged**